File: android/app/src/main/kotlin/com/example/liveactivity/handlers/WorkoutChannelHandler.kt
package com.example.liveactivity.handlers

import android.content.Context
import com.example.liveactivity.helpers.WorkoutServiceHelper
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel

class WorkoutChannelHandler(private val context: Context) {
    companion object {
        const val CHANNEL_NAME = "com.example.healthapp/workoutService"
    }

    fun setUpChannel(dartExecutor: DartExecutor) {
        val channel = MethodChannel(dartExecutor.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "startWorkout" -> {
                    val workoutType = call.argument<String>("workoutType") ?: "Workout"
                    val heartRate = call.argument<Int>("heartRate") ?: 0
                    val steps = call.argument<Int>("steps") ?: 0
                    val calories = call.argument<Int>("calories") ?: 0
                    val elapsedTime = call.argument<Int>("elapsedTime") ?: 0
                    WorkoutServiceHelper.startWorkoutService(context, workoutType, heartRate, steps, calories, elapsedTime)
                    result.success("Workout started")
                }
                "updateWorkout" -> {
                    val workoutType = call.argument<String>("workoutType") ?: "Workout"
                    val heartRate = call.argument<Int>("heartRate") ?: 0
                    val steps = call.argument<Int>("steps") ?: 0
                    val calories = call.argument<Int>("calories") ?: 0
                    val elapsedTime = call.argument<Int>("elapsedTime") ?: 0
                    WorkoutServiceHelper.updateWorkoutService(context, workoutType, heartRate, steps, calories, elapsedTime)
                    result.success("Workout updated")
                }
                "stopWorkout" -> {
                    WorkoutServiceHelper.stopWorkoutService(context)
                    result.success("Workout stopped")
                }
                else -> result.notImplemented()
            }
        }
    }
}

File: ios/Runner/Handler/WorkoutChannelHandler.swift
//
//  WorkoutChannelHandler.swift
//  Runner
//
//  Created by asimsharf on 15/11/2024.
//


import Flutter
import ActivityKit

class WorkoutChannelHandler {
    static let channelName = "com.example.healthapp/workoutService"
    
    static func setUpChannel(with controller: FlutterViewController) {
        let workoutChannel = FlutterMethodChannel(name: channelName, binaryMessenger: controller.binaryMessenger)
        workoutChannel.setMethodCallHandler { call, result in
            switch call.method {
            case "startWorkout":
                if let args = call.arguments as? [String: Any],
                   let workoutType = args["workoutType"] as? String,
                   let goal = args["goal"] as? String {
                    WorkoutActivityService.startWorkoutActivity(workoutType: workoutType, goal: goal)
                    result("Workout started")
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for startWorkout", details: nil))
                }
                
            case "updateWorkout":
                if let args = call.arguments as? [String: Any],
                   let heartRate = args["heartRate"] as? Int,
                   let steps = args["steps"] as? Int,
                   let caloriesBurned = args["caloriesBurned"] as? Int,
                   let elapsedTime = args["elapsedTime"] as? TimeInterval,
                   let goalProgress = args["goalProgress"] as? Double {
                    WorkoutActivityService.updateWorkoutActivity(
                        heartRate: heartRate,
                        steps: steps,
                        caloriesBurned: caloriesBurned,
                        elapsedTime: elapsedTime,
                        goalProgress: goalProgress
                    )
                    result("Workout updated")
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for updateWorkout", details: nil))
                }
                
            case "stopWorkout":
                WorkoutActivityService.stopWorkoutActivity()
                result("Workout stopped")
                
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}
